### PR TITLE
[MRG] Allow asymmetric matrices in plot_connectome

### PR DESCRIPTION
**Pull request by NicolasGensollen**
Addresses #1513 

**Description**

This PR proposes to allow asymmetric matrices to be plotted as a directed graph through the `plot_connectome` function. Previous behaviour was to raise an error.

**Example**

```python
import numpy as np
from nilearn import datasets, plotting
from nilearn.input_data import NiftiLabelsMasker
from nilearn.connectome import ConnectivityMeasure

yeo = datasets.fetch_atlas_yeo_2011()
data = datasets.fetch_development_fmri(n_subjects=10)
connectome_measure = ConnectivityMeasure(kind='correlation')
masker = NiftiLabelsMasker(labels_img=yeo['thick_17'], standardize=True,
                           memory='nilearn_cache')
time_series = []
for func, confounds in zip(data.func, data.confounds):
    time_series.append(masker.fit_transform(func, confounds=confounds))
correlation_matrices = connectome_measure.fit_transform(time_series)
mean_correlation_matrix = connectome_measure.mean_
coordinates = plotting.find_parcellation_cut_coords(labels_img=yeo['thick_17'])
# Brutal way to get asymmetric matrix...
non_symmetric = np.triu(mean_correlation_matrix)
plotting.plot_connectome(mean_correlation_matrix, coordinates,
                         edge_threshold="90%", title='symmetric')
plotting.plot_connectome(non_symmetric, coordinates,
                         edge_threshold="90%", title='asymmetric')
```
directed_connectome

**Todo**

- Add tests (to be coming soon)
- Showcase the functionality in an example (I don't think we have asymmetric connectivity measures implemented in Nilearn so far, so not sure if we can build a realistic example here...)

### CODE REVIEW

**Comment by bthirion** [comment]
LGTM, thx ! 
A possible example to get a directed graph in examples would be to estimate lag-1 correlations between regions (some simplistic Granger causality model).

**Comment by NicolasGensollen** [comment]
Azure failure related to empty arrows crashing for matplotlib pre 3.1 (see: https://github.com/matplotlib/matplotlib/pull/13658)

**Comment by NicolasGensollen** [comment]
>A possible example to get a directed graph in examples would be to estimate lag-1 correlations between regions (some simplistic Granger causality model).

This might be too simplistic, but maybe something along those lines ?

```python
import numpy as np
from nilearn import datasets, plotting
from nilearn.input_data import NiftiLabelsMasker

def lag_correlation(time_series, lag):
    """Basic home-made lag correlation function."""
    n_subjects = len(time_series)
    n_samples, n_features = time_series[0].shape
    lag_cor = np.zeros((n_subjects, n_features, n_features))
    for subject,serie in enumerate(time_series):
        for i in range(n_features):
            for j in range(n_features):
                lag_cor[subject, i, j] = np.corrcoef(serie[lag:, i], 
                                                     serie[:-lag, j])[0, 1]
    return np.mean(lag_cor, axis=0)

# Get atlas and func data
yeo = datasets.fetch_atlas_yeo_2011()
data = datasets.fetch_development_fmri(n_subjects=10)

# Transform into time series
masker = NiftiLabelsMasker(labels_img=yeo['thick_17'], standardize=True,
                           memory='nilearn_cache')
time_series = []
for func, confounds in zip(data.func, data.confounds):
    time_series.append(masker.fit_transform(func, confounds=confounds))

# Compute node coordinates
coordinates = plotting.find_parcellation_cut_coords(labels_img=yeo['thick_17'])

# Compute lag correlation matrices and plot connectomes
for lag in [1, 10]:
    lag_cor  = lag_correlation(time_series, lag)
    plotting.plot_connectome(lag_cor, coordinates, edge_threshold="90%", 
                             title='Lag{} correlation'.format(lag))
```
directed_connectome2

**Comment by bthirion** [comment]
Thx for the example suggestion.
IMHO it does not really make sense to find lag-10 correlations between areas: those are probably noise... I would only take lag-1, and possibly show it together with the symmetric lag-0 correlation.

**Comment by NicolasGensollen** [comment]
>IMHO it does not really make sense to find lag-10 correlations between areas: those are probably noise... I would only take lag-1, and possibly show it together with the symmetric lag-0 correlation.

Thanks for the comment! I agree and I updated one of the connectome-related examples accordingly. 
Here is the link to the modified example built by CircleCI: [link](https://6717-1235740-gh.circle-artifacts.com/0/doc/_build/html/auto_examples/03_connectivity/plot_atlas_comparison.html#sphx-glr-auto-examples-03-connectivity-plot-atlas-comparison-py)

**Comment by bthirion** [comment]
At least it looks OK to me.

File: examples/03_connectivity/plot_atlas_comparison.py [hunk 1]
```diff
@@ -80,6 +80,56 @@
...
+from numpy.testing import assert_array_almost_equal
+
+empirical_connectome_measure = ConnectivityMeasure(
+                                    cov_estimator=EmpiricalCovariance(),
+                                    kind='correlation')
+empirical_correlation_matrices = empirical_connectome_measure.fit_transform(
+                                    time_series)
+mean_empirical_correlation_matrix = empirical_connectome_measure.mean_
+lag0_correlation = lag_correlation(time_series, 0)
+assert_array_almost_equal(mean_empirical_correlation_matrix,
```
Comment (bthirion) [inline examples/03_connectivity/plot_atlas_comparison.py:121]: This is a bit weird in an example.

File: nilearn/plotting/tests/test_img_plotting.py [hunk 2]
```diff
@@ -566,6 +567,58 @@ def test_plot_connectome(tmpdir):
...
+    ax = plot_connectome(adjacency_matrix,
+                         node_coords,
+                         display_mode='lzry')
+    # No edge in direction 'l' because of node coords
+    assert(len([patch for patch in ax.axes['l'].ax.patches
+             if isinstance(patch, FancyArrow)]) == 0)
+    for direction in ['z', 'r', 'y']:
+        assert(len([patch for patch in ax.axes[direction].ax.patches
+             if isinstance(patch, FancyArrow)]) ==
+                    np.prod(adjacency_matrix.shape) - 2)
```
Comment (bthirion) [inline nilearn/plotting/tests/test_img_plotting.py:598]: This is not related to this PR ? I'm confused.

Comment (NicolasGensollen) [reply nilearn/plotting/tests/test_img_plotting.py:598]: I am testing that we get the correct number of arrows in the connectome plot when setting some (2) elements of the matrix to zero and for a different display mode. If this is too redundant with other tests added in this PR, I can delete it

Comment (bthirion) [reply nilearn/plotting/tests/test_img_plotting.py:598]: OK, sounds right.

**Review by bthirion — APPROVED** [review]

**Comment by NicolasGensollen** [comment]
Thanks for the reviews!
Merging.
